src/session.cpp

```diff
@@ -151,83 +129,87 @@ session_t::push_action_t::operator()(std::shared_ptr<channel<tcp>> ptr) {
 
 void
 session_t::push_action_t::finalize(const std::error_code& ec) {
-    if(ec) {
-        session->signals.shutdown(ec);
+    if(ec.value() == 0) return;
+
+    if(ec != asio::error::eof) {
+        COCAINE_LOG_ERROR(session->log, "client disconnected: [%d] %s", ec.value(), ec.message());
+    } else {
+        COCAINE_LOG_DEBUG(session->log, "client disconnected");
     }
-}
 
-class session_t::discard_action_t {
-    synchronized<channel_map_t>& channels;
+    return session->detach(ec);
+}
 
+class session_t::channel_t
+{
 public:
-    discard_action_t(synchronized<channel_map_t>& channels_):
-        channels(channels_)
+    channel_t(const dispatch_ptr_t& dispatch_, const upstream_ptr_t& upstream_):
+        dispatch(dispatch_),
+        upstream(upstream_)
     { }
 
-    void
-    operator()(const std::error_code& ec);
+    dispatch_ptr_t dispatch;
+    upstream_ptr_t upstream;
 };
 
-void
-session_t::discard_action_t::operator()(const std::error_code& ec) {
-    auto ptr = channels.synchronize();
-
-    for(auto it = ptr->begin(); it != ptr->end(); ++it) {
-        if(it->second->dispatch) it->second->dispatch->discard(ec);
-    }
-
-    ptr->clear();
-}
-
 // Session
 
-session_t::session_t(std::unique_ptr<channel<tcp>> transport_, const dispatch_ptr_t& prototype_):
+session_t::session_t(std::unique_ptr<logging::log_t> log_,
+                     std::unique_ptr<channel<tcp>> transport_, const dispatch_ptr_t& prototype_):
+    log(std::move(log_)),
     transport(std::shared_ptr<channel<tcp>>(std::move(transport_))),
     prototype(prototype_),
     max_channel_id(0)
-{
-    signals.shutdown.connect(0, discard_action_t(channels));
-}
+{ }
 
 // Operations
 
 void
 session_t::invoke(const decoder_t::message_type& message) {
-    channel_map_t::const_iterator lb, ub;
-    channel_map_t::key_type channel_id = message.span();
+    const channel_map_t::key_type channel_id = message.span();
 
-    std::shared_ptr<channel_t> channel;
-
-    {
-        auto ptr = channels.synchronize();
+    const auto channel = channels.apply([&](channel_map_t& mapping) -> std::shared_ptr<channel_t> {
+        channel_map_t::const_iterator lb, ub;
 
-        std::tie(lb, ub) = ptr->equal_range(channel_id);
+        std::tie(lb, ub) = mapping.equal_range(channel_id);
 
         if(lb == ub) {
             if(channel_id <= max_channel_id) {
                 // NOTE: Checking whether channel number is always higher than the previous channel
                 // number is similar to an infinite TIME_WAIT timeout for TCP sockets. It might be
-                // not the best approach, but since we have 2^64 possible channels, and it is a lot
-                // more than 2^16 ports for sockets, it is fit to avoid stray messages.
-                return;
+                // not the best approach, but since we have 2^64 possible channels it's good enough.
+                throw cocaine::error_t("specified channel id was revoked");
```

Comment:
Actually, to prevent a situation when a client sends messages to a channel ignoring all errors coming back. Imagine a situation where a client sends `storage::write_streamed(collection, key)` followed by a number of `streaming::write` invocations. In case of the former method actually failing and returning an error, our clueless client would end up calling some completely random method instead of 'streaming::write' since the channel was reset after the error.